README.md
# Racker
[![Gem Version](https://badge.fury.io/rb/racker.svg)](http://badge.fury.io/rb/racker)
[![Build Status](https://travis-ci.org/aspring/racker.svg?branch=master)](https://travis-ci.org/aspring/racker)

* Website: http://github.com/aspring/racker

Racker is an opinionated Ruby DSL for generating Packer(www.packer.io) templates.

Racker is able to take multiple Racker templates, merge them together, and generate a single Packer template.  This process allows for deep merging Packer configuration pieces to give the user a more granular approach for creating and organizing templates, not currently possible with the Packer template format.

## Features

* Allows for building Packer templates from DRY modular templates
* Allows for commenting sections of the template
* Supports use of knockouts when merging

## Installation

    $ gem install racker

## Usage
To generate a Packer template from a Racker template, run:

    $ racker rackertemplate1.rb packertemplate.json

To generate a Packer template from multiple Racker templates, run:

    $ racker rackertemplate1.rb rackertemplate2.rb packertemplate.json

To merge multiple templates you can keep adding Racker templates:

    $ racker rackertemplate1.rb rackertemplate2.rb rackertemplate3.rb packertemplate.json

The output of the previous command will be template 1 deep merged with template2, the result of this merge is deep merged with template 3 then output as a packer template.

## Racker Template Syntax
The goal of Racker is to provide a template structure that allows for allowing full control of the template merging process to achieve the desired Packer template.

In order to do this Racker takes an opinionated stance on the following:

* All arrays within Packer Builder namespace are converted to hashes with well defined keys -- this allows for easy knockout capability based on key name.
* The provisioners hash uses a Fixnum key to allow for defining an order that provisioners will be written to the Packer template.

#### Base Template Syntax

The most basic Racker template would be the following:

```ruby
Racker::Processor.register_template do |t|
end
```

This template would not define a variable, builder, provisioner or post-processor and would be a pretty boring template.

#### Variables Namespace

Racker templates support the `variables` namespace which is a hash.  This hash maps one to one to a Packer template's variables section.

This is an example of a basic `variables` definition:

```ruby
Racker::Processor.register_template do |t|
    # Define the variables
    t.variables = {
        'iso_checksum'              => '6232efa014d9c6798396b63152c4c9a08b279f5e',
        'iso_checksum_type'         => 'sha1',
        'iso_url'                   => 'http://mirrors.kernel.org/centos/6.4/isos/x86_64/CentOS-6.4-x86_64-minimal.iso',
        'kickstart_file'            => 'template1-ks.cfg',
        'vagrant_output_file'       => "./boxes/centos-6.4-{{.Provider}}.box"
    }
end
```

#### Builders Namespace

Racker templates support the `builders` namespace which is a hash, keyed by the name of the builder.

All Packer arrays inside of this namespace should be represented as hashes in Racker.  Racker will use the value when creating the template, the key is there purely for allowing you to override/knockout as necessary.

This is an abbreviated example of adding a builder named 'test' that is a 'virtualbox-iso' builder type:

```ruby
Racker::Processor.register_template do |t|
    # Define the builders
    t.builders['test'] = {
        'type'                      => 'virtualbox-iso',
        'guest_os_type'             => 'RedHat_64',
        'headless'                  => true,
        'format'                    => 'ova',
        'guest_additions_path'      => "VBoxGuestAdditions_{{.Version}}.iso",
        'iso_checksum'              => "{{user `iso_checksum`}}",
        'iso_checksum_type'         => "{{user `iso_checksum_type`}}",
        'iso_url'                   => "{{user `iso_url`}}",
        'virtualbox_version_file'   => '.vbox_version',
        'boot_command'              => {
            0 => '<tab> text ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/{{user `kickstart_file`}}<enter><wait>'
        },
        'boot_wait'                 => '10s',
        'http_directory'            => 'http_directory',
        'ssh_port'                  => 22,
        'ssh_username'              => 'root',
        'ssh_password'              => 'asdfasdf',
        'ssh_wait_timeout'          => '10000s',
        'shutdown_command'          => 'shutdown -P now',
        'disk_size'                 => 8096,
        'vboxmanage'                => {
            'memory'  => [ 'modifyvm', '{{.Name}}', '--memory',    '1024' ],
            'cpus'    => [ 'modifyvm', '{{.Name}}', '--cpus',      '1' ],
            'ioapic'  => [ 'modifyvm', '{{.Name}}', '--ioapic',    'on' ]
        }
    }
end
```
One of the sections of node in this builder is the `vboxmanage`.  It has been converted into a Hash to make it easier to knockout.

#### Provisioners Namespace

Racker templates support the `provisioners` namespace which is a Fixnum keyed hash.

When generating the Packer template, Racker will order the provisioners based upon the Fixnum key, this allows complete control on the ordering of provisioners throughout Racker templates.

```ruby
Racker::Processor.register_template do |t|
    # Define the provisioners
    t.provisioners = {
        0 => {
            'update-cacert-bundle' => {
                'type' => 'shell',
                'script' => 'scripts/common/update-cacert-bundle.sh'
            },
        },
        300 => {
            'install-chef-11.8.0' => {
                'type' => 'shell',
                'script' => 'scripts/common/install-chef-11.8.0.sh'
            },
        },
        500 => {
            'install-guest-additions-dependencies' => {
                'type' => 'shell',
                'scripts' => [
                    'scripts/common/install-guest-additions.sh'
                ],
            },
        },
        750 => {
            'disable-services' => {
                'type' => 'shell',
                'script' => 'scripts/centos-6/disable-services.sh'
            },
        },
        900 => {
            'prepare-vagrant-instance' => {
                'type' => 'shell',
                'script' => 'scripts/common/prepare-vagrant-instance.sh',
                'only' => ['test']
            },
            'prepare-ec2-instance' => {
                'type' => 'shell',
                'script' => 'scripts/common/prepare-ec2-instance.sh',
                'only' => ['amazon']
            },
        }
    }
end
```

#### Post-Processors Namespace

Racker templates support the `postprocessors` namespace which is a hash, keyed by the name of the post-processor.

```ruby
Racker::Processor.register_template do |t|
    # Define the post-processors
    t.postprocessors['vagrant'] = {
        'type' => 'vagrant',
        'output' => '{{user `vagrant_output_file`}}',
        'compression_level' => 7,
        'keep_input_artifact' => true,
        'only' => ['virtualbox-vagrant','vmware-vagrant']
    }
end
```

#### Other keys

You can add `description` or `min_packer_version` to template.

```ruby
Racker::Processor.register_template do |t|
    t.description = 'some description'
    t.min_packer_version = '1.0.0'
end
```
### Putting it all together

Racker offers 2 very basic example templates `example/template1.rb` and `example/template2.rb` as well as the resulting packer template from the two templates run through Racker.

To experiement with these templates, after installing Racker, and cloning this repo you can execute the following:

    $ racker ./example/template1rb ./example/tempalte2.rb ./example/packer.json

While these two templates are not all inclusive of the capabilities of Racker, it shows off the ability to have a basic template, and a second template that removes the pieces of the template that target Amazon, as well as adds two chef solo provisioner steps.

## Testing

Run the tests with `rake`:
```bash
  # Ensure you've installed the bundle:
  bundle install

  # Run spec tests
  bundle exec rake spec

  # Run tests and generate code coverage statistics
  bundle exec rake coverage

  # Run RuboCop assessment
  bundle exec rake rubocop
```

## Outstanding Development

* The following builders have not been fully tested:
    * docker
    * qemu
* Implement unit testing
* Additional documentation work 
* Add capability to target specific packer versions should the packer template format change.
* Add quick init to generate a basic Racker template

## Contributions

Feel free to fork and request a pull, or submit a ticket
https://github.com/aspring/racker/issues

## License

This project is available under the MIT license. See LICENSE for details.


CHANGELOG.md
## Unreleased

* Added support for `min_packer_version` and `description` keys.

## 0.2.0 (2016-03-01)

* Added support for output to STDOUT.  The filename of '-' is used to signify outout should go to STDOUT. (PR#7)  This feature will become the default functionality in the 0.3.0 release.
* Fixed issue where the `parallels` builder was not properly required. 

## 0.1.6 (2014-09-16)

* Fixed `uninitialized constant Racker::Builders::Null` bug.

## 0.1.5 (2014-09-08)

* Added support for VirtualBox `import_flags` options which were added in Packer 0.7.0.

## 0.1.4 (2014-08-07)

New Features:
* Added support for OpenStack `security_groups` options.
* Added `null` builder support.
* Added `parallels` builder support.  (UNTESTED)

## 0.1.3 (2014-05-06)

* Added support for VirtualBox `export_opts` and `vboxmanage_post` options which were added in Packer 0.6.0.

## 0.1.2 (2014-03-11)

* Fix quiet option cli setting

## 0.1.1 (2014-03-06)

New Features:
* Added command line option to surpress output.
* Added the ability to set the knockout prefix from the command line.

Bug Fixes:
* Do not create empty Packer namespace entries if the Racker template does not contain the namespace.
* Changed default knockout prefix from `--` to `~~` as it conflicted with certain Virtualbox builder keys.

## 0.1.0 (2014-02-23)

* Initial public release

## 0.0.3 (Internal)

* Adding support for outstanding builders (digitalocean,docker,google,openstack and qemu)
* Added more robust logging support

## 0.0.2 (Internal)

* Removed ability to name post-processors due to packer template change.

## 0.0.1 (Internal)

* Initial private release
* Supports Amazon, Virtualbox and VMWare builders


example/template1.rb
Racker::Processor.register_template do |t|

  # Define the variables
  t.variables = {
    'iso_checksum'              => '6232efa014d9c6798396b63152c4c9a08b279f5e',
    'iso_checksum_type'         => 'sha1',
    'iso_url'                   => 'http://mirrors.kernel.org/centos/6.4/isos/x86_64/CentOS-6.4-x86_64-minimal.iso',
    'kickstart_file'            => 'template1-ks.cfg',
    'vagrant_output_file'       => "./boxes/centos-6.4-{{.Provider}}.box"
  }

  # Define the builders
  t.builders['virtualbox-vagrant'] = {
    'type'                      => 'virtualbox-iso',
    'guest_os_type'             => 'RedHat_64',
    'headless'                  => true,
    'format'                    => 'ova',
    'guest_additions_path'      => "VBoxGuestAdditions_{{.Version}}.iso",
    'iso_checksum'              => "{{user `iso_checksum`}}",
    'iso_checksum_type'         => "{{user `iso_checksum_type`}}",
    'iso_url'                   => "{{user `iso_url`}}",
    'virtualbox_version_file'   => '.vbox_version',
    'boot_command'              => {
      0 => '<tab> text ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/{{user `kickstart_file`}}<enter><wait>'
    },
    'boot_wait'                 => '10s',
    'http_directory'            => 'http_directory',
    'ssh_port'                  => 22,
    'ssh_username'              => 'root',
    'ssh_password'              => 'asdfasdf',
    'ssh_wait_timeout'          => '10000s',
    'shutdown_command'          => 'shutdown -P now',
    'disk_size'                 => 8096,
    'vboxmanage'                => {
      'memory'  => [ 'modifyvm', '{{.Name}}', '--memory',    '1024' ],
      'cpus'    => [ 'modifyvm', '{{.Name}}', '--cpus',      '1' ],
      'ioapic'  => [ 'modifyvm', '{{.Name}}', '--ioapic',    'on' ]
    }
  }

  t.builders['vmware-vagrant'] = {
    'type'                      => 'vmware-iso',
    'guest_os_type'             => 'centos-64',
    'headless'                  => true,
    'tools_upload_flavor'       => 'linux',
    'iso_checksum'              => '{{user `iso_checksum`}}',
    'iso_checksum_type'         => '{{user `iso_checksum_type`}}',
    'iso_url'                   => '{{user `iso_url`}}',
    'boot_command'              => {
      0 => '<tab> text ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/{{user `kickstart_file`}}<enter><wait>'
    },
    'boot_wait'                 => '10s',
    'http_directory'            => 'http_directory',
    'ssh_port'                  => 22,
    'ssh_username'              => 'root',
    'ssh_password'              => 'asdfasdf',
    'ssh_wait_timeout'          => '10000s',
    'shutdown_command'          => 'shutdown -P now',
    'disk_size'                 => 8096,
    'vmdk_name'                 => 'centos',
    'vmx_data'                  => {
      'cpuid.coresPerSocket'          => '1',
      'memsize'                       => '1024',
      'numvcpus'                      => '1',
      'isolation.tools.copy.disable'  => 'false',
      'isolation.tools.paste.disable' => 'false'
    }
  }

  t.builders['amazon'] = {
    'type'                      => 'amazon-ebs',
    'region'                    => 'us-east-1',
    'source_ami'                => 'ami-testami',
    'instance_type'             => 't1.micro',
    'ssh_username'              => 'root',
    'ssh_timeout'               => '5m',
    'ami_name'                  => 'packer-centos-6 {{timestamp}}',
    'ami_block_device_mappings' => {
      '/dev/sda' => {
        'device_name'           => '/dev/sda',
        'volume_size'           => 8,
        'delete_on_termination' => true
      },
      '/dev/sdb' => {
        'device_name'           => '/dev/sdb',
        'volume_size'           => 8,
        'delete_on_termination' => true
      }
    },
    'ami_regions' => [
      'us-west-1',
      'us-west-2'
    ]
  }

  # Define the provisioners
  t.provisioners = {
    0 => {
      'yum-install-packages' => {
        'type' => 'shell',
        'script' => 'scripts/centos-6/yum-install-packages.sh'
      },
      'disable-sshd-dns-lookup' => {
        'type' => 'shell',
        'script' => 'scripts/common/disable-sshd-dns-lookup.sh'
      },
      'no-tty-sudo' => {
        'type' => 'shell',
        'script' => 'scripts/common/no-tty-sudo.sh'
      },
      'set-xen-device-name' => {
        'type' => 'shell',
        'script' => 'scripts/common/set-xen-device-names.sh'
      },
      'update-cacert-bundle' => {
        'type' => 'shell',
        'script' => 'scripts/common/update-cacert-bundle.sh'
      },
    },
    300 => {
      'install-chef-11.8.0' => {
        'type' => 'shell',
        'script' => 'scripts/common/install-chef-11.8.0.sh'
      },
    },
    500 => {
      'install-guest-additions-dependencies' => {
        'type' => 'shell',
        'scripts' => [
          'scripts/centos-6/install-guest-additions-dependencies.sh',
          'scripts/common/install-guest-additions.sh'
        ],
        'only' => ['virtualbox-vagrant', 'vmware-vagrant']
      },
    },
    750 => {
      'disable-services' => {
        'type' => 'shell',
        'script' => 'scripts/centos-6/disable-services.sh'
      },
    },
    900 => {
      'prepare-vagrant-instance' => {
        'type' => 'shell',
        'script' => 'scripts/common/prepare-vagrant-instance.sh',
        'only' => ['virtualbox-vagrant']
      },
      'prepare-ec2-instance' => {
        'type' => 'shell',
        'script' => 'scripts/common/prepare-ec2-instance.sh',
        'only' => ['amazon']
      },
      'update-staic-ip-files' => {
        'type' => 'file',
        'source' => 'uploads/static-ip',
        'destination' => '/root',
        'only' => ['vmware-vagrant']
      },
    },
    999 => {
      'yum-remove-packages' => {
        'type'    => 'shell',
        'script'  => 'scripts/centos-6/yum-remove-packages.sh'
      },
      'clean-unnecessary-files' => {
        'type'    => 'shell',
        'script'  => 'scripts/centos-6/clean-unnecessary-files.sh'
      },
      'zero-empty-space' => {
        'type'    => 'shell',
        'script'  => 'scripts/common/zero-empty-space.sh'
      },
    }
  }

  # Define the post-processors
  t.postprocessors['vagrant'] = {
    'type' => 'vagrant',
    'output' => '{{user `vagrant_output_file`}}',
    'compression_level' => 7,
    'keep_input_artifact' => true,
    'only' => ['virtualbox-vagrant','vmware-vagrant']
  }

end
